### ISSUE TITLE
TypeError issue: exceptions must derive from BaseException  in a multiple apps when dealing with json

### ISSUE THREAD

**Issue by chasx003** [frikky/Shuffle#775]
**Describe the bug**

Ran force-update of all apps sometime 3/24 or 3/25, and afterwards multiple apps are failing with TypeError issue: exceptions must derive from BaseException  --- Seeing in Hive create case app where json is passed for custom objects, and can also reproduce in shuffle app

**To Reproduce**
1. Force App Update
2. Create merge list shuffle tool node
3. List 1  =  [ { "thing": "object1" }, { "thing": "object2" }, { "thing": "object3" } ]
4. list 2 = [ { "thing2": "true" }, { "thing2": "True" }, { "thing": "true" } ]
5. Run - Creates TypeError issue: exceptions must derive from BaseException

https://cdn.discordapp.com/attachments/747813673577218063/957082777143103549/unknown.png
...

**Expected behavior**
Previous versions would not have the error and merge lists correctly

**Debug logs (NOT APPLICABLE FOR CLOUD)**
https://pastebin.com/fcZ9yE89

**Comment by 00willo** [comment]
The underlying cause of the TypeError listed in the output is
`Got exec error: the JSON object must be str, bytes or bytearray, not list`

I tested and that actual problem is in both the 1.1.0 and 1.0.0 versions of the app.

Version 1.0.0:
```
$ docker run -it --entrypoint=/bin/sh -it frikky/shuffle:shuffle-tools_1.0.0 
/app # python
Python 3.10.0 (default, Oct 26 2021, 23:20:47) [GCC 10.3.1 20210424] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import logging
>>> from app import Tools
>>> logger = logging.getLogger(__name__)
>>> first_list = [{'thing': 'object1'}, {'thing': 'object2'}, {'thing': 'object3'}]
>>> second_list = [{'thing2': 'true'}, {'thing2': 'True'}, {'thing': 'true'}]
>>> foo = Tools(None, logger)
>>> foo.merge_lists(first_list, second_list)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/app/app.py", line 1018, in merge_lists
    list_one = json.loads(list_one)
  File "/usr/local/lib/python3.10/json/__init__.py", line 339, in loads
    raise TypeError(f'the JSON object must be str, bytes or bytearray, '
TypeError: the JSON object must be str, bytes or bytearray, not list
```

Version 1.1.0:
```
$ docker run -it --entrypoint=/bin/sh -it frikky/shuffle:shuffle-tools_1.1.0
/app # python
Python 3.10.0 (default, Oct 26 2021, 23:20:47) [GCC 10.3.1 20210424] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import logging
>>> from app import Tools
>>> logger = logging.getLogger(__name__)
>>> first_list = [{'thing': 'object1'}, {'thing': 'object2'}, {'thing': 'object3'}]       
>>> second_list = [{'thing2': 'true'}, {'thing2': 'True'}, {'thing': 'true'}]
>>> foo = Tools(None, logger)
>>> foo.merge_lists(first_list, second_list)  
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/app/app.py", line 1303, in merge_lists
    list_one = json.loads(list_one)
  File "/usr/local/lib/python3.10/json/__init__.py", line 339, in loads
    raise TypeError(f'the JSON object must be str, bytes or bytearray, '
TypeError: the JSON object must be str, bytes or bytearray, not list
```

diff_lists also experiences the same problem.
```
>>> foo.diff_lists(first_list, second_list)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/app/app.py", line 1002, in diff_lists
    list_one = json.loads(list_one)
  File "/usr/local/lib/python3.10/json/__init__.py", line 339, in loads
    raise TypeError(f'the JSON object must be str, bytes or bytearray, '
TypeError: the JSON object must be str, bytes or bytearray, not list
```

Further looking at the `merge_lists` method called by the instance of the `Tools` class, I can see exactly why the TypeError around the handling of the `list_one` argument is occuring.
```
    def merge_lists(self, list_one, list_two, set_field="", sort_key_list_one="", sort_key_list_two=""):
        try:
            list_one = json.loads(list_one)
        except json.decoder.JSONDecodeError as e:
            print("Failed to parse list1 as json: %s" % e)
```

Based on the output of the "unit" style test case I used above and the handling of the inputs, my assumptions about the structure of the input data to the method is not valid.  They're not straight lists, it's a list embeded in a json dict looking thing.

Not exactly sure why it's different (when inputting via the UI) from before and after doing a force reload of the apps from the default github repo.  Something inbetween the data input into the UI paramaters and what gets sent as the input to these functions has changed.

Investigation continues.

**Comment by 00willo** [comment]
It's looking likely, that there's a couple fixes required for this:

- some further input validation on these methods
- the handling of the execption returned (the `exceptions must derive from BaseException` part from the original output log)
- and, whatever has caused the change in the input data. TDB

**Comment by frikky** [comment]
We've indeed seen these a lot in the last few days. I do think it has to do with an ast.literal_eval() we added to the SDK, which was supposed to fix broken JSON objects.

This is overall not a "bad" thing, as correct types are showing up - the problem is just that it is not consistent with old versions. May roll it back if we can't find a clear fix.

https://github.com/frikky/Shuffle/blob/cff2b888451a1173cce43e8fd0886a25898923b3/backend/app_sdk/app_base.py#L3054

**Comment by 00willo** [comment]
I did see the `Failed parsing value with ast:` appearing in the container logs from that try/except block, but in some cases it didn't appear to be causing a failure.

But, in only starting to dig in a few days ago, I didn't know if it was anything to worry about yet. 😄 

I started to look at the `app_base.py`, and quickly got lost in attempting to reason the what and why's of what was occurring. There are some big methods in there, hahaha.

**Comment by frikky** [comment]
> I did see the `Failed parsing value with ast:` appearing in the container logs from that try/except block, but in some cases it didn't appear to be causing a failure.
> 
> But, in only starting to dig in a few days ago, I didn't know if it was anything to worry about yet. smile
> 
> I started to look at the `app_base.py`, and quickly got lost in attempting to reason the what and why's of what was occurring. There are some big methods in there, hahaha.

Hehe, I think we'll have to visualize that the app_base.py at some point to make it clear in what order what happens. There are so many ways to run with it, and parse things that it's hard if you're not getting properly introduced to it.

I made one hotfix to it just now. As the issue seems to be parsing things that shouldn't be, we're now forcing it to only look at strings that could potentially be JSON (which was the original point): https://github.com/frikky/Shuffle/commit/3d5e02d8dabf4c81d5c947e8bd780465c3f0819f

@chasx003 could you try pulling the SDK and force updating again please?

**Comment by chasx003** [comment]
> 

Can you provide instructions on how to update the SDK? I must be doing something wrong.  I stopped all shuffle containers, purged pretty much all the images on the box that had to do anything with shuffle / tagged frikky, restarted the stack and rebuilt apps, but I can see the app sdk still doesn't have the newest commit you're referencing

**Comment by frikky** [comment]
Two steps!
1. docker pull frikky/shuffle:app_sdk
2. Click force update in Shuffle (/apps top right)

**Comment by chasx003** [comment]
> Two steps!
> 
> 1. docker pull frikky/shuffle:app_sdk
> 2. Click force update in Shuffle (/apps top right)

Ah, I actually did that before the image prune... Just did the same now, and still getting same results, app_sdk shows up to date, but still doesn't have the new commits

**Comment by 00willo** [comment]
I was able to upate mine, by just doing the force update. 

But, I still get the same result as well.

```
2022-03-26 22:59:09,796 - Tools - INFO:[DEBUG] Got exec error: the JSON object must be str, bytes or bytearray, not list
2022-03-26 22:59:09,795 - Tools - INFO:[INFO] Running normal execution (not loop)
2022-03-26 22:59:09,796 - Tools - INFO:TypeError issue: exceptions must derive from BaseException
```

```
$ docker history frikky/shuffle:shuffle-tools_1.1.0
IMAGE          CREATED        CREATED BY                                      SIZE      COMMENT
cb9937a02bc9   9 hours ago    /bin/sh -c #(nop)  CMD ["python" "app.py" "-…   0B        
fb999f7559d3   9 hours ago    /bin/sh -c #(nop) WORKDIR /app                  0B        
cce38001c1aa   9 hours ago    /bin/sh -c apk --no-cache add jq git curl       1.4MB     
383da1ee587f   9 hours ago    /bin/sh -c #(nop) COPY dir:629c8dc6720a684bd…   65.7kB    
4a89d1e88d40   9 hours ago    /bin/sh -c #(nop) COPY dir:e9d2da173e3c8eafc…   75.3MB    
42115e5b3cde   12 hours ago   /bin/sh -c #(nop) COPY file:d8ddaea6c6cc83d8…   158kB     
<missing>      13 days ago    /bin/sh -c #(nop) COPY file:136ad7dce53261b4…   0B        
<missing>      13 days ago    /bin/sh -c pip3 install -r /requirements.txt    25.2MB    
<missing>      13 days ago    /bin/sh -c #(nop) COPY file:44aadd084a3a7348…   117B      
<missing>      4 months ago   /bin/sh -c #(nop) COPY dir:17f063a4bec351fbf…   0B        
<missing>      4 months ago   /bin/sh -c apk update && apk add --update tz…   220MB     
<missing>      5 months ago   /bin/sh -c #(nop)  CMD ["python3"]              0B        
<missing>      5 months ago   /bin/sh -c set -ex;   wget -O get-pip.py "$P…   8.31MB    
<missing>      5 months ago   /bin/sh -c #(nop)  ENV PYTHON_GET_PIP_SHA256…   0B        
<missing>      5 months ago   /bin/sh -c #(nop)  ENV PYTHON_GET_PIP_URL=ht…   0B        
<missing>      5 months ago   /bin/sh -c #(nop)  ENV PYTHON_SETUPTOOLS_VER…   0B        
<missing>      5 months ago   /bin/sh -c #(nop)  ENV PYTHON_PIP_VERSION=21…   0B        
<missing>      5 months ago   /bin/sh -c cd /usr/local/bin  && ln -s idle3…   32B       
<missing>      5 months ago   /bin/sh -c set -ex  && apk add --no-cache --…   29.8MB    
<missing>      5 months ago   /bin/sh -c #(nop)  ENV PYTHON_VERSION=3.10.0    0B        
<missing>      7 months ago   /bin/sh -c #(nop)  ENV GPG_KEY=A035C8C19219B…   0B        
<missing>      7 months ago   /bin/sh -c set -eux;  apk add --no-cache   c…   1.76MB    
<missing>      7 months ago   /bin/sh -c #(nop)  ENV LANG=C.UTF-8             0B        
<missing>      7 months ago   /bin/sh -c #(nop)  ENV PATH=/usr/local/bin:/…   0B        
<missing>      7 months ago   /bin/sh -c #(nop)  CMD ["/bin/sh"]              0B        
<missing>      7 months ago   /bin/sh -c #(nop) ADD file:aad4290d27580cc1a…   5.6MB 
```


```
$ docker run -it --entrypoint=/bin/sh -it frikky/shuffle:shuffle-tools_1.1.0

/app # sed -n "3049,3060p" walkoff_app_sdk/app_base.py

                            try:
                                for key, value in params.items():
                                    try:
                                        if isinstance(value, str) and ((value.startswith("{") and value.endswith("}")) or (value.startswith("[") and value.endswith("]"))):
                                            params[key] = ast.literal_eval(value)
                                    except Exception as e:
                                        self.logger.info(f"[DEBUG] Failed parsing value with ast: {e}")
                                        continue
                            except Exception as e:
                                self.logger.info("[DEBUG] Failed looping objects. Non critical: {e}")
```
It looks like, I've definitely got the change in my base image.

**Comment by chasx003** [comment]
Yeaaah, Its one of those days.  Mine was updated, for some reason I was being silly and thinking the change was in the ast.literal_eval line, forgot it was the "and" above it.  So I have the same results as well, still failing

**Comment by frikky** [comment]
Gonna dig into this properly now. Hope to find a proper fix.

**Comment by frikky** [comment]
After some manual testing, it looks like it's on the apps side, as they're expecting to have to parse the lists themselves. On top of this, our SDK didn't handle the errors properly, making the exception very generic and hard to understand.

I'll go over the main apps and handle this properly today. Looks like it will be a long-term good thing, but possibly cause some errors short-term.

**Comment by frikky** [comment]
So lots of issues have been fixed, and we'll be releasing a proper fix for it all later tonight, as this week has had too many breaking changes.

Here's a couple:
- ast.literal_eval will stay, and has shown that some parts of our error handling is bad. This has been fixed.
- Our testing systems weren't good enough. We've now built another workflow that tests every single action whether they work before deploying. We'll be making this one public as well, as it can be used for examples too.

**Comment by frikky** [comment]
This has indeed been fixed! To get the latest fixes; update apps:
```
#1 
docker pull frikky/shuffle:app_sdk

#2 
# Go to /apps and force update apps.
```

If required: restart Orborus to run the latest containers

**Comment by 00willo** [comment]
@chasx003 I was able to confirm that the error for your actual issue is now reflective of the cause. 

![Screen Shot 2022-03-28 at 8 10 19 pm](https://user-images.githubusercontent.com/6183357/160367772-e97506f4-6940-4c66-964c-b462072b02cc.png)

 @frikky mentioned in https://github.com/Shuffle/python-apps/issues/283#issuecomment-1079951463 that they're still working on a v1.2.0 release of Shuffle-Tools, which should fix this list merge issue.

I confirmed the following input now results in a more meaningful exception.

```
list_one = [ { "thing": "object1" }, { "thing": "object2" }, { "thing": "object3" } ]
list_two = [ { "thing2": "true" }, { "thing2": "True" }, { "thing": "true" } ]
```

```
list_one = [1, 2, 3]
list_two = [4, 5, 6]
```

They all currently result in `the JSON object must be str, bytes or bytearray, not list` exception, this appears to be occuring from the `json.loads` of the input lists in the shuffle tools app.  @frikky just mentioned in the discord chat that they're working on fixing it today.
